packages/core/src/createActor.ts

```diff
@@ -397,6 +406,22 @@ export class Actor<TLogic extends AnyActorLogic>
     };
   }
 
+  public on(
+    emittedEventType: string,
+    handler: EmittedEventHandler
+  ): Subscription {
+    const set = this.eventListeners.get(emittedEventType) ?? new Set();
+    set.add(handler);
```

Comment:
You have to wrap the input handler because otherwise you end up with weird inconsistencies:
```ts
actorRef.on('foo', handler)
const sub = actorRef.on('foo', handler)

sub.unsubscribe()

emit({ type: 'foo' })

// radio silence 😢
```

Reply:
Are you sure? `sub.unsubscribe()` is basically the same as a `.removeEventListener(handler)`. 

The radio silence seems like the correct behavior.

Reply:
The API is quite different here from the `removeEventListener` API though. With `addEventListener`/`removeEventListener` you don't get a subscription object back - so your `listener` acts as a cache key in the internal set. For that reason, if you add the same listener twice it will be called once and if you remove it once then it will remove all of the subscribers (well, the only one that was added). The actual storage algorithm is slightly more complicated because it also accounts for options - u might add the same listener with different options and IIRC that would create separate internal entries. That's beside the point though :P

With the subscription API, the situation is different because you get back a subscription object - the intuition is that this object is bound to a specific *instance* of the subscription.

This is how it always worked with RxJS and most of the other stream libraries ([stackblitz](https://stackblitz.com/edit/rxjs-uvymmo?devtoolsheight=60&file=index.ts)):
```ts
import { Subject } from 'rxjs';

const subject = new Subject();

function listener(v: unknown) {
  console.log('got', v);
}

const first = subject.subscribe({
  next: listener,
});
subject.subscribe({
  next: listener,
});

first.unsubscribe();

subject.next('hello!'); // it still gets logged
```

This is also how our own `.subscribe` API works, how Redux works (see [here](https://github.com/reduxjs/redux/blob/dfcfab2d9a399c3fe19e12a47607ced44acda6b5/src/createStore.ts#L223-L224)) and how TanStack Query v4 worked (see [here](https://github.com/TanStack/query/blob/bdc746abdbf40df809648b1501240391a06e833d/packages/query-core/src/subscribable.ts#L11-L13)). Interestingly, TanStack Query v5 has changed this behavior (see [here](https://github.com/TanStack/query/blob/55782aebbbc69020858f8ed611ab41d24eee215a/packages/query-core/src/subscribable.ts#L11-L12)).

I've also tried to take a look at the current proposed `Observable` API in Chrome Canary (spec [here](https://github.com/WICG/observable)). However... that just doesn't support `.unsubscribe` at all and each subscription just sets up a new subscription each time:
```ts
var stream$ = new Observable((s) => {
  console.log("setup");
  setTimeout(() => {
    s.next("hey!"), s.complete();
  }, 1);
});

function listener(v) {
  console.log("got", v);
}

stream$.subscribe({ next: listener });
stream$.subscribe({ next: listener });
// logs both `setup` and `got` twice
```

So this isn't exactly comparable because with this `Observable` spec we create two separate sources whereas our API is about subscribing to the same source. "Sharing" could be implemented here though - whatever you do in your setup callback is up to you, so  - in there - you could reach into a shared resource and setup ur own sharing logic etc. That sharing logic wouldn't even be able to "deduplicate" those subscriptions because the callback doesn't get any access to the user-provided callbacks. It just gets access to its own unique subscription object and only that can be used as "cache key" etc.

If we implement teardown on top of the example above we can see how it only cancels a single subscriber here:
```ts
const controller = new AbortController();

let id = 0;

const stream$ = new Observable((s) => {
  const myId = ++id;
  const timerId = setTimeout(() => {
    s.next(`hey from ${myId}!`);
  }, 1);

  s.addTeardown(() => {
    console.log("tearing down", myId);
    clearTimeout(timerIt);
  });
});

function listener(v) {
  console.log("got", v);
}

stream$.subscribe({ next: listener }, { signal: controller.signal });
stream$.subscribe({ next: listener });

controller.abort(); // this would only cancel the first one!
```

Reply:
Makes sense; I'll wrap it then 👍 🍬